third-party/rsutils/src/network-adapter-watcher.cpp

```diff
@@ -0,0 +1,234 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2024 Intel Corporation. All Rights Reserved.
+
+#include <rsutils/os/network-adapter-watcher.h>
+#include <rsutils/shared-ptr-singleton.h>
+#include <rsutils/signal.h>
+#include <rsutils/easylogging/easyloggingpp.h>
+
+#if ! defined( __APPLE__ ) && ! defined( __ANDROID__ )
+#ifdef _WIN32
+
+#include <windows.h>
+#include <iphlpapi.h>
+#pragma comment( lib, "iphlpapi.lib" )
+
+#else
+
+#include <sys/socket.h>
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#include <sys/eventfd.h>
+#include <unistd.h>
+#include <poll.h>
+
+#endif
+#endif  // ! __APPLE__ && ! __ANDROID__
+
+
+namespace rsutils {
+namespace os {
+namespace detail {
+
+
+#if ! defined( __APPLE__ ) && ! defined( __ANDROID__ )
+class network_adapter_watcher_singleton
+{
+    using public_signal = rsutils::public_signal< network_adapter_watcher_singleton >;
+
+#ifdef _WIN32
+    HANDLE _done;
+#else
+    int _socket;
+    int _done;
+#endif
+    std::thread _th;
+
+public:
+    public_signal callbacks;
+
+    network_adapter_watcher_singleton()
+#ifdef _WIN32
+        : _done( CreateEvent(
+            nullptr /*security attrs*/, TRUE /*manual reset*/, FALSE /*init unset*/, nullptr /*no name*/ ) )
+        , _th(
+              [this]
+              {
+                  // We're going to wait on the following handles:
+                  HANDLE handles[2];
+                  DWORD const n_handles = sizeof( handles ) / sizeof( handles[0] );
+                  handles[0] = _done;
+                  // https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-notifyaddrchange
+                  OVERLAPPED overlap = { 0 };
+                  overlap.hEvent = CreateEvent( nullptr /*security attrs*/,
+                                                FALSE /*auto reset*/,
+                                                FALSE /*init unset*/,
+                                                nullptr /*no name*/ );
+                  handles[1] = overlap.hEvent;
+
+                  LOG_DEBUG( "starting network adapter watcher" );
+
+                  while( true )
+                  {
+                      HANDLE h_file = NULL;
+                      NotifyAddrChange( &h_file, &overlap );
+                      // If multiple objects become signaled, the function returns the index of the first handle in the
+                      // array whose object was signaled:
+                      if( WAIT_OBJECT_0
+                          == WaitForMultipleObjects( n_handles, handles, FALSE /*wait-for-all*/, INFINITE ) )
+                          break;  // we're done
+                      LOG_DEBUG( "network adapter changes detected!" );
+                      callbacks.raise();
```

Comment:
Can also reach here on wait failures.
Consider if needs to specifically check for WAIT_OBJECT_1.

Reply:
I can add code looking for `WAIT_FAILED` but this is really not expected

Reply:
I guess I'll make it stop waiting, too...

Reply:
Done